src/vos/vos_dtx.c

```diff
@@ -1350,6 +1350,19 @@ vos_dtx_check_availability(struct umem_instance *umm, daos_handle_t coh,
 
 		/* PUNCH DTX cannot be shared by others. */
 		if (dtx->te_intent == DAOS_INTENT_PUNCH) {
+			if (dth == NULL)
+				/* XXX: For rebuild case, if some normal IO
+				 *	has generated punch-record (by race)
+				 *	before rebuild logic handling that,
+				 *	then rebuild logic should ignore such
+				 *	punch-record, because the punch epoch
+				 *	will be higher than the rebuild epoch.
+				 *	The rebuild logic needs to create the
+				 *	original target record that exists on
+				 *	other healthy replicas before punch.
+				 */
+				return ALB_UNAVAILABLE;
```

Comment:
I saw this patch but sorry I'm not quite understand the logic.
here the (dth == NULL) seems it is possible for enumerate or fetch operation, and find there is an existed related PREPARED punch DTX? seems it is not only for rebuild? (normal enumerate/fetch is not related with this path?).

Reply:
For fetch and enumerate cases, the intent will be DAOS_INTENT_DEFAULT, that has been handled before moving here, at line 1328-1329.

Here, if the @dth is NULL, then it is a rebuild update.

Reply:
thanks for explanation.